testing/sync-test/src/auth.rs

```diff
@@ -165,19 +165,55 @@ impl TestAccount {
             &client_id,
             &scope,
             &jwk,
-            &self.auth_key,
+            &auth_key,
             &self.cfg,
             (&self.k_sync, &self.xcs),
         )?;
-        let auth_params = auth::AuthorizationParameters::new(
-            client_id.clone(),
-            code_challenge.clone(),
-            code_challenge_method.clone(),
-            scope.clone(),
-            keys_jwe,
-            state.clone(),
-        );
-        auth::send_authorization_request(&self.cfg, auth_params, &self.auth_key)
+        let auth_params = auth::AuthorizationRequestParameters {
+            client_id: client_id.clone(),
+            code_challenge: Some(code_challenge.clone()),
+            code_challenge_method: Some(code_challenge_method.clone()),
+            scope: scope.clone(),
+            keys_jwe: Some(keys_jwe),
+            state: state.clone(),
+            access_type: "offline".to_string(),
+        };
+        auth::send_authorization_request(&self.cfg, auth_params, &auth_key)
+    }
+
+    fn execute_oauth_pair_flow(&self, oauth_uri: &str) -> Result<(String, String)> {
+        let url = Url::parse(&oauth_uri)?;
+        let query_map: HashMap<String, String> = url.query_pairs().into_owned().collect();
+        let keys_jwk = query_map.get("keys_jwk").unwrap();
+        let scope = query_map.get("scope").unwrap();
+        let client_id = query_map.get("client_id").unwrap();
+        let state = query_map.get("state").unwrap();
+        let code_challenge = query_map.get("code_challenge").unwrap();
+        let code_challenge_method = query_map.get("code_challenge_method").unwrap();
+        let scoped_keys = auth::get_scoped_keys(
+            scope,
+            client_id,
+            &auth::derive_auth_key_from_session_token(&self.session_token)?,
+            &self.cfg,
+            (&self.k_sync, &self.xcs),
+        )?;
```

Comment:
Wow there is a lot of unpacking code here! Not a blocker, but I wonder if this would be easier to follow if you implemented `From<Url>` for `AuthorizationParameters` in order to factor out the unpacking into a separate code unit.